enhancements/dns/coredns-egressfirewall-integration.md

```diff
@@ -0,0 +1,363 @@
+* The OVN-K master will delete the ACL rule(s) corresponding to the EgressFirewall rule(s) containing the wildcard DNS name(s).
+* The OVN-K master will then check if the same wildcard DNS names are also used in the EgressFirewall rules in other Namespaces. If
+they are not used, then the OVN-K master will delete the corrresponding ``AdressSet`` for each of the DNS names in the EgressFirewall
+rules. The OVN-K master will also delete the corresponding ``DNSName`` CRs, only if the ``.spec.isregular`` field is set to false and
+the ``.spec.iswildcard`` field is set to true. If both the fields are set to true, then the CR will not be deleted and the
+``.spec.iswildcard`` field will be set to false. From the ``.status`` field all the other DNS names' details will be removed and
+only the details of the DNS name will be kept which matches with the Name field of the CR.
+
+
+#### Variation [optional]
+
+
+### API Extensions
+
+The validation of ``DNSName`` field in ``EgressFirewallDestination`` will be updated to accept wildcard DNS names as well.
+It will be updated from ``^([A-Za-z0-9-]+\.)*[A-Za-z0-9-]+\.?$`` which accepts only regular DNS names to
+``^(\*\.)?([A-Za-z0-9-]+\.)*[A-Za-z0-9-]+\.?$``.
+
+````go
+// EgressFirewallDestination is the endpoint that traffic is either allowed or denied to
+type EgressFirewallDestination struct {
+	// ..
+
+	// dnsName is the domain name to allow/deny traffic to. If this is set, cidrSelector must be unset.
+	// +kubebuilder:validation:Pattern=^(\*\.)?([A-Za-z0-9-]+\.)*[A-Za-z0-9-]+\.?$
+	DNSName string `json:"dnsName,omitempty"`
+}
+````
+
+The following ``DNSName`` CRD will be added to the ``k8s.ovn.org`` api-group.
+
+````go
+// DNSName describes a DNS name used in a EgressFirewall rule.
+type DNSName struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Specification of the desired behavior of DNSName.
+	Spec DNSNameSpec `json:"spec"`
+	// Observed status of DNSName
+	// +optional
+	Status DNSNameStatus `json:"status,omitempty"`
+}
+
+// DNSNameSpec is a desired state description of DNSName.
+type DNSNameSpec struct {
+	// IsRegular denotes whether the resource is representing a regular DNS name.
+	// Cannot be false if IsWildcard is also false.
+	// +optional
+	IsRegular bool `json:"isregular,omitempty"`
+	// IsWildcard denotes whether the resource is representing a wildcar DNS name.
+	// Cannot be false if IsRegular is also false.
+	// +optional
+	IsWildcard bool `json:"iswildcard,omitempty"`
+}
+
+type DNSNameStatus struct {
+	// The list of matching DNS names and their corresponding IPs along with TTL and next
+	// time to DNS lookup.
+	Items []DNSNameStatusItem `json:"items"`
+}
+
+type DNSNameStatusItem struct {
+	DNSName string        `json:"dnsname"`
+	Info    []DNSNameInfo `json:"info"`
+}
+
+type DNSNameInfo struct {
+	IP             string `json:"ip"`
+	TTL            string `json:"ttl"`
+	NextLookupTime string `json:"nextlookuptime"`
```

Comment:
Typically we include a `lastUpdated` rather than `nextUpdate`, you can calculate the next lookup by using the TTL, but it's also useful to see when was it last updated, for example, are you expecting the update to happen only after the data expires? I would probably expect the logic to see if we are close to the TTL based on when it was last updated and update it _before_ it expires right? The current API makes me think it's being updated after it expires

Reply:
Makes sense. I will use `lastUpdated` instead of `NextLookupTime`.